cmd/lncli/psbt.go

```diff
@@ -0,0 +1,283 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/signal"
+	"github.com/urfave/cli"
+)
+
+// openChannelPSBT starts an interactive channel open protocol that uses a
+// partially signed bitcoin transaction (PSBT) to fund the channel output. The
+// protocol involves several steps between the RPC server and the CLI client:
+//
+// RPC server                           CLI client
+//     |<----register funding shim----------|
+//     |                                    |
+//     |  |<------open channel (stream)-----|
+//     |  |-------ready for funding----->|  |
+//     |  |<------PSBT verify------------|  |
+//     |  |-------ready for signing----->|  |
+//     |  |<------PSBT finalize----------|  |
+//     |  |-------channel pending------->|  |
+//     |  |-------channel open------------->|
+//     |                                    |
+func openChannelPSBT(ctx *cli.Context, client lnrpc.LightningClient,
+	req *lnrpc.OpenChannelRequest, block bool) error {
+
+	var (
+		pendingChanID [32]byte
+		shimPending   = true
+		quit          = make(chan struct{})
+		ctxc, cancel  = context.WithCancel(context.Background())
+	)
+	defer cancel()
+	if _, err := rand.Read(pendingChanID[:]); err != nil {
+		return fmt.Errorf("unable to generate random chan ID: %v", err)
+	}
+	fmt.Printf("Starting PSBT funding flow with pending channel ID %x\n",
+		pendingChanID)
+
+	// Create the PSBT funding shim and register it with the wallet.
+	req.FundingShim = &lnrpc.FundingShim{
+		Shim: &lnrpc.FundingShim_PsbtShim{
+			PsbtShim: &lnrpc.PsbtShim{
+				PendingChanId: pendingChanID[:],
```

Comment:
Don't think we need to make them specify the pending channel ID here. Instead things can go directly to using open to kick things off.